### PR TITLE
DAOS-13012 test: Adding test_replay_check_pointing test

### PR DESCRIPTION
**Pull request by phender**
Adding test to verify data access after engine restart with check pointing.

Skip-unit-tests: true
Test-tag: ReplayTests
Test-nvme: auto_md_on_ssd

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by phender** [comment]
Test passed https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-12007/1/artifact/Functional%20Hardware%20Medium/server/replay.py/results.html

File: src/tests/ftest/server/replay.py [hunk 1]
```diff
@@ -369,3 +370,53 @@ def test_replay_no_check_pointing(self):
...
+                frequency = entry['value']
+        self.log.info('%s check point frequency: %s seconds', container.pool, frequency)
+        if frequency == 0:
+            self.fail('Error obtaining the pool check point frequency')
+
+        self.log_step('Write data to the container (ior)')
+        ior = self.write_data(container, ppn)
+
+        self.log_step('Waiting for check pointing to complete (sleep {})'.format(frequency))
+        time.sleep(frequency)
```
Comment (rpadma2) [inline src/tests/ftest/server/replay.py:415]: How do we know checkpointing is complete within this time? Any flag check not available?

Comment (phender) [reply src/tests/ftest/server/replay.py:415]: This was the proposed implementation summarized by https://daosio.atlassian.net/browse/DAOS-13016?focusedCommentId=114023.

File: src/tests/ftest/server/replay.py [hunk 2]
```diff
@@ -369,3 +374,40 @@ def test_replay_no_check_pointing(self):
...
+        frequency = 5
+        container = self.create_container(
+            properties=f'checkpoint:timed,checkpoint_freq:{frequency}')
+        self.log.info('%s check point frequency: %s seconds', container.pool, frequency)
+
+        self.log_step('Write data to the container (ior)')
+        ior = self.write_data(container, ppn)
+
+        self.log_step('Waiting for check pointing to complete (sleep {})'.format(frequency))
+        time.sleep(frequency)
```
Comment (wangshilong) [inline src/tests/ftest/server/replay.py:406]: maybe time.sleep(frequency + 1) or time.sleep(frequency *2) looks more safe.

Comment (phender) [reply src/tests/ftest/server/replay.py]: Done.

File: src/tests/ftest/server/replay.py [hunk 3]
```diff
@@ -369,3 +370,53 @@ def test_replay_no_check_pointing(self):
...
+            if entry['name'] == 'checkpoint' and entry['value'] != 'timed':
+                self.log_step(
+                    'Setting the {} check pointing to timed (dmg pool set-prop)'.format(
+                        container.pool))
+                container.pool.set_prop(properties='checkpoint:timed')
+            elif entry['name'] == 'checkpoint_freq':
+                frequency = entry['value']
+        self.log.info('%s check point frequency: %s seconds', container.pool, frequency)
+        if frequency == 0:
+            self.fail('Error obtaining the pool check point frequency')
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/server/replay.py]: As discussed, the test now explicitly sets the checkout style and frequency.

**Review by rpadma2 — APPROVED** [review]

**Review by daltonbohning — APPROVED** [review]

**Review by jolivier23 — COMMENTED** [review]
Looks fine, I think. I wonder if we should add some pool properties for the WAL usage?  @NiuYawei what do you think?   I was thinking the yesterday this would be useful for myself to just do some touch testing.   Is it too much information.

**Comment by NiuYawei** [comment]
> Looks fine, I think. I wonder if we should add some pool properties for the WAL usage? @NiuYawei what do you think? I was thinking the yesterday this would be useful for myself to just do some touch testing. Is it too much information.

Could elaborate on what kind of pool properties? I always have concerns of exporting too much information to user, so if it's supposed being valuable for developer only, maybe we'd implement it through env variable or fault injection etc.?

**Comment by phender** [comment]
All tests passed in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-12007/6/testReport/FTEST_server/ReplayTests/

**Comment by phender** [comment]
We should land this after https://github.com/daos-stack/daos/pull/11797 lands as it will be easier to rebase this PR.

**Comment by phender** [comment]
Also confirmed that the fix for populating the server config file with the bdev_roles is working, e.g. https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-12007/6/artifact/Functional%20Hardware%20Medium/server/replay.py/daos_configs.wolf-110/daos_server.yml/*view*/

**Pull request merged**